middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -515,18 +527,20 @@ let split_complex_binding ~env ~res (binding : complex binding) =
         effs;
         inline = binding.inline;
         bound_expr = Split { cmm_expr = new_cmm_expr };
-        cmm_var = binding.cmm_var
+        cmm_var = binding.cmm_var;
+        machtype =
+          P.Without_args.result_kind prim |> To_cmm_utils.machtype_of_kind
       }
     in
     res, Split { new_bindings; split_binding }
 
 (* Adding binding to the env and split them *)
 
-let rec add_binding_to_env ?extra env res var (Binding binding as b) =
+let rec add_binding_to_env ?extra env res var machtype (Binding binding as b) =
   let env =
     let bindings = Variable.Map.add var b env.bindings in
     let cmm_expr = C.var (Backend_var.With_provenance.var binding.cmm_var) in
-    let vars = Variable.Map.add var cmm_expr env.vars in
+    let vars = Variable.Map.add var (cmm_expr, machtype) env.vars in
```

Comment:
You should be able to get the machtype from the binding instead of requiring it as an explicit argument.